openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -1,22 +1,180 @@
+            )
+        except (AttributeError, IndexError, ValueError) as error:
+            raise PathstreamTaskExecutionError(
+                '_prepare_new_content',
+                'An enrollment data is invalid. Reason: {} \nNew data: {}'.format(str(error), content)
+            )
+
+        # Convert elements of content to bytes.
+        content = map(lambda data: data.encode(), content)
+
+        return b''.join(content)
+
+    def _upload_file(self, file_content=None):
+        """Upload the file to an S3 bucket using MD5 to ensure that data is not corrupted
+        traversing the network.
+
+        :param: file_content (bytes)."""
+        if not (isinstance(file_content, bytes) and file_content):
+            raise PathstreamTaskExecutionError(
+                '_upload_file', 'Can\'t upload a file with empty content or invalid content type.'
+            )
+
+        hash_algorithm_md5 = hashlib.md5(file_content).digest()
+        content_encoded = base64.b64encode(hash_algorithm_md5)
+
+        try:
+            response = self.client.put_object(
+                Bucket=self.S3_BUCKET,
+                Key=self.S3_FILE,
+                Body=file_content,
+                ContentMD5=content_encoded.decode(),  # String - The base64-encoded 128-bit MD5 digest of the message.
+            )
+        except ClientError as error:
+            raise PathstreamTaskExecutionError(
+                '_upload_file',
+                'Failed to upload the file to S3. Reason: {}'.format(str(error))
+            )
+
+        LOG.info('File [%s] was uploaded to S3 for [%s], response: [%s]', self.S3_FILE, self.__str__(), response)
+
+    def execute_upload(self):
+        """
+        Context: This method will download, update and upload the S3 file content with the
+        corresponding data from new enrollments which have not been uploaded to the S3
+        file. It also updates the key flag 'is_uploaded' which is used to determine if an enrollment is already
+        uploaded to the S3 file.
+
+        :return: tuple (boolean, str). (True, 'successful-message') if the method executes properly,
+        otherwise (False, 'error-message')
+
+        True means that this method either download-update-upload the S3 file content or do not
+        proceed to download-update-upload the content because there are no new enrollments.
+        """
+        users_enrollments_qs = ExternalEnrollment.objects.filter(  # pylint: disable=no-member
+            controller_name=str(self),
+            meta__icontains='"is_uploaded":false',
+        )
+
+        if not users_enrollments_qs:
+            LOG.info('There are no new enrollments to update the S3 file.')
+
+            return COMPLETED, 'execute_upload completed'
+
+        new_content = []
+
+        for user_enrollments in users_enrollments_qs:
+            for enrollment in user_enrollments.meta:
+                if not enrollment.get('is_uploaded'):
```

Comment:
This block will add the new content even if 'is_uploaded' is not present or is empty... @Jacatove

Reply:
Good catch!. Now it uses a default value to avoid that case. @Squirrel18